### PR TITLE
Add Path Template and loaders to handle dynamic Pages

### PR DESCRIPTION
**Pull request by igorbrasileiro**
## Context
The Live handler only fetches pages from the storage with `url.pathname` that's fixed values. A problem exists when you have dynamic routes based on a fresh framework, because the storage should have each URL that matches the dynamic route.

## Solution
To tackle this problem, the createLivePage receives options parameter with interface `CreateLivePageOptions` and behaves following this pattern:
- if options is empty, the Handler runs default loader that get ComponentData from CMS and passes the loader output to the `ctx.render`.
- if options have a template, then Handler runs default loader considering the `template` and `url.pathname` and passes the loader output to the `ctx.render`.
- if options have a loader, the Handler runs options.loader instead of default loader and passes loader output to the `ctx.render`.

### Usage examples:
```tsx
// routes/static.tsx
// static page with loader

import { createLiveHandler, LivePage } from "$live/live.tsx";

export const handler = createLiveHandler();
export default LivePage;
```

```tsx
// routes/[slug]/p.tsx
// Custom Render and dynamic page

import { createLiveHandler, LiveComponents } from "$live/live.tsx";
import Layout from "../components/Layout.tsx";

export const handler = createLiveHandler({ template: "/:slug/p" });
export default function MyCustomRoute({ data }: PageProps<LivePageData>) {
  return (
    <Layout>
      <LiveComponents components={data.components} />
    </Layout>
  );
}
```

```tsx
// routes/static.tsx
// Custom render and custom loader
import {
  createLiveHandler,
  LiveComponents,
  loadLiveComponents,
} from "$live/live.tsx";
import Layout from "../components/Layout.tsx";

export interface MyLoaderData {
  userId: number;
  id: number;
  title: string;
  completed: boolean;
}

const loadTodos = (req, ctx) =>
  fetch("https://jsonplaceholder.typicode.com/todos/1").then((res) =>
    res.json()
  );

const loader = async (req, ctx) => {
  const [todos, components] = await Promise.all([
    loadTodos(req, ctx),
    loadLiveComponents(req, ctx),
  ]);
  return { todos, components };
};

export const handler = createLiveHandler({ loader });

export default function MyCustomRoute({ data }: PageProps<LivePageData>) {
  return (
    <Layout>
      <LiveComponents components={data.components} />
    </Layout>
  );
}
```

### CODE REVIEW

File: live.tsx [hunk 1]
```diff
@@ -91,22 +104,25 @@ export function createLivePage<Data>(options: LivePageOptions<Data>) {
...
-      : <div>Page not found</div>;
-    const InspectVSCode = !isDenoDeploy &&
+    const { components, loaderData } = data;
+    const renderComponents =
+      components && components.length > 0 ? (
+        components.map(({ component, props }: any) => {
+          const Comp =
+            manifest.islands[`./islands/${component}.tsx`]?.default ||
+            manifest.components![`./components/${component}.tsx`]?.default;
+          return <Comp {...props} loaderData={loaderData} />;
```
Comment (igorbrasileiro) [inline live.tsx:117]: Should pass loaderData to components?

**Review by lucis — APPROVED** [review]

File: live.tsx [hunk 2]
```diff
@@ -91,22 +104,25 @@ export function createLivePage<Data>(options: LivePageOptions<Data>) {
...
+        })
+      ) : defaultRender ? (
+        defaultRender(props)
+      ) : (
+        <div>Page not found</div>
+      );
```
[the remark this answers is not in the document]
Comment (igorbrasileiro) [reply live.tsx:123]: Isso é resultado do meu lint, não mudei esse código, tirando a parte do loaderData.

File: live.tsx [hunk 3]
```diff
@@ -59,24 +55,41 @@ export function createLivePage<Data>(options: LivePageOptions<Data>) {
         return new Response("Site not found", { status: 404 });
       }
 
-      let { data: Pages, error } = await getSupabaseClient()
+      /**
+       * Queries follow PostgREST syntax
+       * https://postgrest.org/en/stable/api.html#horizontal-filtering-rows
+       */
+      const queries = [url.pathname, template]
```
Comment (guifromrio) [inline live.tsx:62]: Beautiful

File: live.tsx [hunk 4]
```diff
@@ -59,24 +55,41 @@ export function createLivePage<Data>(options: LivePageOptions<Data>) {
...
+        .filter((query) => Boolean(query))
+        .map((query) => `path.eq.${query}`)
+        .join(",");
+
+      const pagePromise = getSupabaseClient()
         .from("Pages")
         .select(`components, path, site!inner(name, id)`)
         .eq("site.name", site)
-        .eq("path", url.pathname);
+        .or(queries);
```
[the remark this answers is not in the document]
Comment (guifromrio) [reply live.tsx:71]: Ou bate exato ou bate template

[the remark this answers is not in the document]
Comment (igorbrasileiro) [reply live.tsx:71]: Um OR, por que ou dá match na rota fixa ou no template.

Ex: tenho /:slug/p ou /café/p
ele consegue pegar os dois casos, por isso o OR

File: live.tsx (hunk 1 above)
Comment (guifromrio) [reply live.tsx:117]: Well, for now, yes. It's straightforward and clear.

File: live.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (guifromrio) [reply live.tsx:123]: Alinhado com proposta do Lucis

**Review by guifromrio — APPROVED** [review]

File: live.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (igorbrasileiro) [reply live.tsx:123]: O que acha de fazermos isso em outra PR?

[the remark this answers is not in the document]
Comment (igorbrasileiro) [reply live.tsx]: Corrigi

File: live.tsx [hunk 5]
```diff
@@ -19,22 +19,66 @@ export const setupLive = (manifest: DecoManifest, liveOptions: LiveOptions) => {
...
+): Promise<LivePageData> {
+  const site = userOptions.site;
+  const url = new URL(req.url);
+  const { template } = options ?? {};
+
+  /**
+   * Queries follow PostgREST syntax
+   * https://postgrest.org/en/stable/api.html#horizontal-filtering-rows
+   */
+  const queries = [url.pathname, template]
```
Comment (guifromrio) [inline live.tsx:49]: will this behave nicely if empty? perform 2 queries?

**Review by guifromrio — APPROVED** [review]
LGTM!

File: live.tsx (hunk 5 above)
Comment (igorbrasileiro) [reply live.tsx:49]: Nope, if template is undefined, only perform 1 query page = url.pathname.

**Pull request merged**